packages/vscode-extension/src/debug/taskTerminal/baseTaskTerminal.ts

```diff
@@ -0,0 +1,47 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as vscode from "vscode";
+
+const ControlCodes = {
+  CtrlC: "\u0003",
+};
+
+export abstract class BaseTaskTerminal implements vscode.Pseudoterminal {
+  protected writeEmitter = new vscode.EventEmitter<string>();
+  onDidWrite: vscode.Event<string> = this.writeEmitter.event;
+  protected closeEmitter = new vscode.EventEmitter<number>();
+  onDidClose?: vscode.Event<number> = this.closeEmitter.event;
+
+  constructor(private taskDefinition: vscode.TaskDefinition) {}
+
+  open(): void {
+    this.do()
+      .then(() => this.stop())
+      .catch((error) => this.stop(error));
+  }
+
+  close(): void {
+    this.stop();
+  }
+
+  handleInput(data: string): void {
+    this.writeEmitter.fire(data === "\r" ? "\r\n" : data);
+    if (data.includes(ControlCodes.CtrlC)) {
+      this.stop();
+    }
+  }
+
+  // TODO: end local debug session
```

Comment:
Prefer end local debug session in `onDidEndTaskProcessHandler`, and make this pure for terminal.
Is it possible? I remember we already have such logic in `onDidEndTaskProcessHandler`.

Reply:
```
export async function validateLocalPrerequisitesHandler(): Promise<string | undefined> {
  const additionalProperties: { [key: string]: string } = {};
  {
    // If we know this session is concurrently running with another session, send that correlationId in `debug-all-start` event.
    // Mostly, this happens when user stops debugging while preLaunchTasks are running and immediately hit F5 again.
    const session = commonUtils.getLocalDebugSession();
    if (session.id !== commonUtils.DebugNoSessionId) {
      additionalProperties[TelemetryProperty.DebugConcurrentCorrelationId] = session.id;
      // Indicates in which stage (of the first F5) the user hits F5 again.
      additionalProperties[TelemetryProperty.DebugConcurrentLastEventName] =
        localTelemetryReporter.getLastEventName();
    }
  }
  return await Correlator.runWithId(commonUtils.startLocalDebugSession(), async () => {
    if (commonUtils.checkAndSkipDebugging()) {
      // return non-zero value to let task "exit ${command:xxx}" to exit
      return "1";
    }

    await sendDebugAllStartEvent(additionalProperties);
    const result = await localPrerequisites.checkAndInstall();
    if (result.isErr()) {
      // Only local debug use validate-local-prerequisites command
      await sendDebugAllEvent(result.error);
      commonUtils.endLocalDebugSession();
      // return non-zero value to let task "exit ${command:xxx}" to exit
      return "1";
    }
  });
}
```

Now each command will call  `commonUtils.endLocalDebugSession();` when it failed. It is related to the telemetry.

Reply:
So is this `TODO` already done?

Reply:
In `onDidEndTaskProcessHandler`, check the task result, if exit code is non-zero, then call `terminateAllRunningTeamsfxTasks`. So agree to make this class pure for the task logic only.

Reply:
> So is this `TODO` already done?

no, only vscode command will call commonUtils.endLocalDebugSession(). After we moved them to task, may be it still need to be called.

Reply:
May be we can redesign the local debug session in commonUtils and call `commonUtils.endLocalDebugSession()` in `onDidEndTaskProcessHandler`

Reply:
`onDidEndTaskProcessHandler` already calls `terminateAllRunningTeamsfxTasks()`. Is that enough?
I remember VSC stops debug if dependent task has error, thus we only need to stop other running tasks.

Reply:
Here the debug session means the local debug session used to generate session id and send telemetry. As we discussed offline, for teamsfx tasks in a teamsfx project, we need to add the following:
1. start the debug session in `onDidStartTaskProcessHandler` if the exsiting debug session is undefined.
2. stop the debug session in `onDidEndTaskProcessHandler` for non-background task
3. stop the debug session in `onDidStartDebugSessionHandler` for background task (exsiting)